File: go/leetcode/101. Symmetric Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    if root == nil {return true}
    return check2(root.Left, root.Right)
}

func check2(left, right *TreeNode)  bool {
    if left == nil && right == nil {return true}
    if left == nil || right == nil {return false}
    return left.Val == right.Val && 
    check2(left.Right, right.Left) &&
    check2(left.Left, right.Right)
} 
 

File: c/leetcode/101. Symmetric Tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

bool check2(struct TreeNode *left, struct TreeNode *right)
{
  if (!left && !right)
    return true;
  if (!left || !right)
    return false;
  return left->val == right->val &&
         check2(left->right, right->left) &&
         check2(left->left, right->right);
}

bool isSymmetric(struct TreeNode *root)
{
  if (!root)
    return true;
  return check2(root->left, root->right);
}
